lib/Optimizer/Transforms/LinearCtrlRelations.cpp

```diff
@@ -0,0 +1,147 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_LINEARCTRLRELATIONS
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+#define DEBUG_TYPE "linear-ctrl-relations"
+
+using namespace mlir;
+
+namespace {
+class ThreadControl : public OpRewritePattern<quake::ToControlOp> {
+public:
+  explicit ThreadControl(MLIRContext *ctx, DominanceInfo &di)
+      : OpRewritePattern(ctx), dom(di) {}
+
+  LogicalResult matchAndRewrite(quake::ToControlOp toCtrl,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<Operation *> users(toCtrl->getUsers().begin(),
+                                   toCtrl->getUsers().end());
+    auto numUsers = users.size();
+
+    // Check the trivial cases.
+    if (numUsers < 1)
+      return failure();
+    if (numUsers == 1) {
+      // User must be a FromControlOp. We can erase them both.
+      if (auto fromCtrl =
+              dyn_cast_or_null<quake::FromControlOp>(users.front())) {
+        fromCtrl.replaceAllUsesWith(toCtrl.getQubit());
+        rewriter.eraseOp(fromCtrl);
+        rewriter.eraseOp(toCtrl);
+        return success();
+      }
+      // The IR must be broken.
+      return failure();
+    }
+
+    // Now, we can rethread the more interesting case when a value of type
+    // !quake.control is used in actual operations.
+    //
+    // 1. Get a list of the users in dominance order. We have to thread the
+    // !quake.wire value respecting the dominance of the users.
+    SmallVector<Operation *> orderedUsers(numUsers);
+    for (auto *user : users) {
+      if (!user)
+        continue;
+      if ([&]() {
+            for (auto iter = orderedUsers.begin(), iterEnd = orderedUsers.end();
+                 iter != iterEnd; ++iter) {
+              if (!*iter)
+                continue;
```

Comment:
Same question as lines 60-61

Reply:
This may have been extra paranoid from the explosion at the bottom of the cliff. Let me see.